spec/src/main/java/io/a2a/json/JsonUtil.java

```diff
@@ -0,0 +1,729 @@
+                };
+            } catch (IllegalArgumentException e) {
+                throw new JsonSyntaxException("Invalid Part.Kind: " + kindString, e);
+            }
+        }
+    }
+
+    /**
+     * Gson TypeAdapter for serializing and deserializing {@link Part} and its subclasses.
+     * <p>
+     * This adapter handles polymorphic deserialization based on the "kind" field, creating the
+     * appropriate subclass instance (TextPart, FilePart, or DataPart).
+     * <p>
+     * The adapter uses a two-pass approach: first reads the JSON as a tree to inspect the "kind"
+     * field, then deserializes to the appropriate concrete type.
+     *
+     * @see Part
+     * @see TextPart
+     * @see FilePart
+     * @see DataPart
+     */
+    static class PartTypeAdapter extends TypeAdapter<Part<?>> {
+        // Create separate Gson instance without the Part adapter to avoid recursion
+        private final Gson delegateGson = new GsonBuilder()
+                .registerTypeAdapter(OffsetDateTime.class, new OffsetDateTimeTypeAdapter())
+                .registerTypeAdapter(TaskState.class, new TaskStateTypeAdapter())
+                .registerTypeAdapter(Message.Role.class, new RoleTypeAdapter())
+                .registerTypeAdapter(Part.Kind.class, new PartKindTypeAdapter())
+                .registerTypeHierarchyAdapter(FileContent.class, new FileContentTypeAdapter())
+                .create();
+
+        @Override
+        public void write(JsonWriter out, Part<?> value) throws java.io.IOException {
+            if (value == null) {
+                out.nullValue();
+                return;
+            }
+            // Delegate to Gson's default serialization for the concrete type
+            if (value instanceof TextPart textPart) {
+                delegateGson.toJson(textPart, TextPart.class, out);
+            } else if (value instanceof FilePart filePart) {
+                delegateGson.toJson(filePart, FilePart.class, out);
+            } else if (value instanceof DataPart dataPart) {
+                delegateGson.toJson(dataPart, DataPart.class, out);
+            } else {
+                throw new JsonSyntaxException("Unknown Part subclass: " + value.getClass().getName());
+            }
+        }
+
+        @Override
+        public @Nullable Part<?> read(JsonReader in) throws java.io.IOException {
+            if (in.peek() == com.google.gson.stream.JsonToken.NULL) {
+                in.nextNull();
+                return null;
+            }
+
+            // Read the JSON as a tree so we can inspect the "kind" field
+            com.google.gson.JsonElement jsonElement = com.google.gson.JsonParser.parseReader(in);
+            if (!jsonElement.isJsonObject()) {
+                throw new JsonSyntaxException("Part must be a JSON object");
+            }
+
+            com.google.gson.JsonObject jsonObject = jsonElement.getAsJsonObject();
+            com.google.gson.JsonElement kindElement = jsonObject.get("kind");
```

Comment:
When do we have a `kind` field in JSON? 
That's no longer part of the 1.0 spec afaict